Mathlib/Tactic/Simproc/Matrix.lean

```diff
@@ -0,0 +1,135 @@
+/-
+Copyright (c) 2025 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import Mathlib.Data.Matrix.Reflection
+
+/-!
+# Simprocs for matrices
+
+In this file we provide a simproc and a "rw proc" for transpose of explicit matrices.
+
+Existing theorems are `Matrix.cons_transpose` and a simproc `Matrix.cons_val`, but this simproc
+and rw proc aim to do it in one step to optimise the proof term and intermediate goal sizes.
+-/
+
+open Matrix
+open Lean Meta Simp Simproc Tactic Qq
+
+attribute [-simp] cons_transpose
+
+namespace Qq
+
+/-- Decompose a vector expression into a vector of expressions. -/
+def piFinLit? {u : Level} {n : ℕ} {R : Q(Type u)} (e : Q(Fin $n → $R)) :
+    MetaM (Option (Fin n → Q($R))) := do
+  match n with
+  | 0 => return .some ![]
+  | _+1 =>
+    let ~q(vecCons $h $et) := e | return .none
+    let .some t ← piFinLit? et | return .none
+    return .some (vecCons h t)
+
+/-- Decompose a "double vector" expression, i.e. `Fin m → Fin n → R`, into a double vector of
+expressions. -/
+def piFinPiFinLit? {u : Level} {m n : ℕ} {R : Q(Type u)} (e : Q(Fin $m → Fin $n → $R)) :
+    MetaM (Option (Fin m → Fin n → Q($R))) := do
+  match m with
```

Comment:
I think there's a shorter implementation here which is something like `(piFinLit? h).map piFinLit?`

Reply:
Though probably yours is more performant

Reply:
I'm new to monads and tactics, and I don't know how to make (the correct versions of) `map` work with three monads here (`MetaM`, `Option`, and `Fin n -> \.`)